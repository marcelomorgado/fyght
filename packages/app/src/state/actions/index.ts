export * from "./myFyghter";
export * from "./enemies";
export * from "./messages";
export * from "./metamask";
export * from "./daiBalances";
