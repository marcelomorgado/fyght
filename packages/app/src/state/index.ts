import { createStore, createHook, createContainer } from "react-sweet-state";
import {
  fetchAllEnemies,
  initializeMetamask,
  setMetamaskAccount,
  fetchMyFyghter,
  challengeAnEnemy,
  changeMyFyghterSkin,
  createFyghter,
  renameMyFyghter,
} from "./actions";

type State = FyghtContext;

const initialState: State = {
  myFyghter: null,
  enemies: [],
  messages: { errorMessage: null, infoMessage: null },
  metamask: {
    networkId: null,
    account: null,
    ethereum: null,
    contracts: { fyghters: null, dai: null },
    provider: null,
    loading: true,
  },
};

const actions = {
  fetchAllEnemies,
  initializeMetamask,
  setMetamaskAccount,
  fetchMyFyghter,
  challengeAnEnemy,
  changeMyFyghterSkin,
  createFyghter,
  renameMyFyghter,
};

type Actions = typeof actions;

const Store = createStore<State, Actions>({
  initialState,
  actions,
});

export const useFyghtState = createHook(Store);
const FyghtStateContainer = createContainer(Store);
