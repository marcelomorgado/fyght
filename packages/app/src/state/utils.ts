/* eslint-disable no-undef */
import { TransactionReceipt } from "ethers/providers";
import { Transaction } from "ethers/utils";

const ETHEREUM_NETWORK_ID = process.env.ETHEREUM_NETWORK_ID;

// TODO: onFinally or just finally function
export const optimisticUpdate = async ({
  doTransaction,
  onOptimistic,
  onSuccess,
  onError,
  getState,
}: {
  doTransaction: () => Promise<Transaction>;
  onOptimistic?: () => void;
  onSuccess?: (receipt?: TransactionReceipt) => void;
  onError: (errorMessage: string, receipt?: TransactionReceipt) => void;
  getState: () => FyghtState;
}): Promise<void> => {
  const {
    metamask: { loomProvider, ethereumProvider },
  } = getState();

  if (onOptimistic) {
    onOptimistic();
  }

  try {
    const tx = await doTransaction();
    const { chainId } = tx;

    console.log(tx);

    const provider = `${chainId}` === ETHEREUM_NETWORK_ID ? ethereumProvider : loomProvider;

    provider.once(tx.hash, (receipt: TransactionReceipt) => {
      const { status } = receipt;
      if (!status) {
        onError("", receipt);
        return;
      }
      if (onSuccess) {
        onSuccess(receipt);
      }
    });
  } catch (e) {
    // Note: Keeping for now due debug purposes
    console.log(e);

    const errorMessage =
      e.data && e.data.message
        ? e.data.message.replace("VM Exception while processing transaction: revert ", "")
        : "Unexpected error";

    onError(errorMessage);
  }
};
