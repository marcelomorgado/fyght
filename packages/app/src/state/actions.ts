import { StoreActionApi } from "react-sweet-state";
import { ethers, Event, ContractReceipt, BigNumber } from "ethers";
import { optimisticUpdate } from "./utils";
import { Skin, MIN_DEPOSIT } from "../constants";
const { getAddress } = ethers.utils;

type State = FyghtContext;
type StoreApi = StoreActionApi<State>;

export const fetchMyFyghter = () => async ({ setState, getState }: StoreApi): Promise<void> => {
  const {
    metamask: {
      contracts: { fyghters },
      account,
    },
  } = getState();

  const filter = fyghters.filters.FyghterCreated(getAddress(account), null, null);
  const logs = await fyghters.queryFilter(filter, 0, "latest");
  const [myFyghterId] = logs.map((l: Event) => l.args).map(({ id }: FyghterCreated) => id);

  if (myFyghterId) {
    const myFyghter = await fyghters.fyghters(myFyghterId);
    setState({ myFyghter });
  }
};

export const fetchAllEnemies = () => async ({ setState, getState }: StoreApi) => {
  const {
    metamask: {
      contracts: { fyghters },
      provider,
      account,
    },
    myFyghter,
  } = getState();

  if (!fyghters || !provider) {
    setState({ enemies: [] });
    return;
  }

  const loadEnemy = async (id: BigNumber): Promise<Enemy> => {
    const fyghter: Fyghter = await fyghters.fyghters(id);

    let winProbability = null;

    if (myFyghter && myFyghter.id) {
      const { id: myFyghterId } = myFyghter;
      winProbability = await fyghters.calculateWinProbability(myFyghterId, id);
    }
    return { fyghter, winProbability };
  };

  const filter = fyghters.filters.FyghterCreated(null, null, null);
  const logs = await fyghters.queryFilter(filter, 0, "latest");

  const enemiesIds = logs
    .map((l: Event) => l.args)
    .filter(({ owner }: FyghterCreated) => getAddress(owner) !== getAddress(account))
    .map(({ id }: FyghterCreated) => id);

  const enemiesPromises = enemiesIds.map((id: BigNumber) => loadEnemy(id));
  const enemies: Enemy[] = await Promise.all(enemiesPromises);

  setState({ enemies });
};

export const setErrorMessage = (errorMessage: string) => ({ setState, getState }: StoreApi): void => {
  const { messages } = getState();
  setState({ messages: { ...messages, errorMessage } });
};

export const setInfoMessage = (infoMessage: string) => ({ setState, getState }: StoreApi): void => {
  const { messages } = getState();
  setState({ messages: { ...messages, infoMessage } });
};

export const renameMyFyghter = (newName: string) => async ({ getState, setState }: StoreApi): Promise<void> => {
  const {
    myFyghter,
    metamask: {
      contracts: { fyghters },
    },
  } = getState();
  const { id: myFyghterId, name: oldName } = myFyghter;

  optimisticUpdate({
    doTransaction: () => fyghters.rename(myFyghterId, newName),
    onOptimistic: () => {
      setState({ myFyghter: { ...myFyghter, name: newName } });
    },
    onError: (errorMessage: string) => {
      setState({ myFyghter: { ...myFyghter, name: oldName } });
      setErrorMessage(errorMessage);
    },
    getState,
  });
};

export const createFyghter = (name: string) => async ({ getState, setState, dispatch }: StoreApi): Promise<void> => {
  const {
    metamask: {
      contracts: { fyghters, dai },
    },
  } = getState();

  optimisticUpdate({
    doTransaction: async () => {
      // TODO: Create a button for this? (Refs: https://github.com/marcelomorgado/fyght/issues/118)
      await dai.mint(MIN_DEPOSIT);
      // TODO: Call for approval only if needed (Refs: https://github.com/marcelomorgado/fyght/issues/118)
      await dai.approve(fyghters.address, MIN_DEPOSIT);
      return fyghters.create(name);
    },
    onOptimistic: () => {
      const myFyghter: Fyghter = {
        id: null,
        skin: Skin.NAKED,
        name,
        xp: BigNumber.from("1"),
        balance: BigNumber.from(MIN_DEPOSIT),
      };
      setState({ myFyghter });
    },
    onSuccess: async () => {
      dispatch(fetchMyFyghter());
    },
    onError: (errorMessage: string) => {
      setState({ myFyghter: null });
      setErrorMessage(errorMessage);
    },
    getState,
  });
};

export const changeMyFyghterSkin = (newSkin: string) => async ({ getState, setState }: StoreApi): Promise<void> => {
  const {
    myFyghter,
    metamask: {
      contracts: { fyghters },
    },
  } = getState();
  const { id: myFyghterId, skin: oldSkin } = myFyghter;

  optimisticUpdate({
    doTransaction: () => fyghters.changeSkin(myFyghterId, newSkin),
    onOptimistic: () => {
      setState({ myFyghter: { ...myFyghter, skin: newSkin } });
    },
    onError: (errorMessage: string) => {
      setState({ myFyghter: { ...myFyghter, skin: oldSkin } });
      setErrorMessage(errorMessage);
    },
    getState,
  });
};

export const challengeAnEnemy = (enemyId: BigNumber, whenFinish: () => void) => async ({
  getState,
}: StoreApi): Promise<void> => {
  const {
    myFyghter: { id: myFyghterId },
    metamask: {
      contracts: { fyghters },
    },
  } = getState();

  optimisticUpdate({
    doTransaction: () => fyghters.challenge(myFyghterId, enemyId),
    onSuccess: (receipt: ContractReceipt) => {
      const [log] = receipt.logs
        .map((log: Event) => fyghters.interface.parseLog(log))
        .filter(({ name }) => name == "ChallengeOccurred")
        .map(({ args }) => args);
      const [myFyghterId, , winnerId] = log;
      if (winnerId.eq(myFyghterId)) {
        setInfoMessage("You won!");
      } else {
        setInfoMessage("You lose!");
      }

      fetchMyFyghter();
      // TODO: fetch only the enemy
      fetchAllEnemies();
      whenFinish();
    },
    onError: (errorMessage: string) => {
      setErrorMessage(errorMessage);
      whenFinish();
    },
    getState,
  });
};

const setMetamaskNetworkId = (networkId: number) => ({ setState, getState }: StoreApi) => {
  const { metamask } = getState();
  setState({ metamask: { ...metamask, networkId } });
};

export const setMetamaskAccount = (account: string) => ({ setState, getState }: StoreApi) => {
  const { metamask } = getState();
  setState({ metamask: { ...metamask, account } });
};

export const initializeMetamask = () => async ({ setState, getState, dispatch }: StoreApi) => {
  const { metamask } = getState();

  const { ethereum } = window;
  if (ethereum) {
    ethereum.autoRefreshOnNetworkChange = false;
  }
  // Note: Parcel doesn't support process.env es6 destructuring
  //

  // Refs:
  // https://github.com/parcel-bundler/parcel/issues/2299#issuecomment-439768971
  // https://en.parceljs.org/env.html

  // eslint-disable-next-line no-undef
  const FYGHTERS_CONTRACT_ADDRESS = process.env.FYGHTERS_CONTRACT_ADDRESS;
  // eslint-disable-next-line no-undef
  const DAI_CONTRACT_ADDRESS = process.env.DAI_CONTRACT_ADDRESS;

  if (ethereum) {
    // Note: The metamask docs recommends to use the 'chainChanged' event instead but it isn't working
    // See more: https://docs.metamask.io/guide/ethereum-provider.html#methods-new-api
    ethereum.on("networkChanged", (networkId: number) => {
      dispatch(setMetamaskNetworkId(networkId));
    });

    ethereum.on("accountsChanged", ([account]: string[]) => {
      dispatch(setMetamaskAccount(account));
    });

    const provider = new ethers.providers.Web3Provider(ethereum);
    const signer = provider.getSigner();
    const FYGHTERS_CONTRACT_ABI = require("../contracts/Fyghters.json").abi;
    const DAI_CONTRACT_ABI = require("../contracts/Dai.json").abi;

    const fyghters = new ethers.Contract(FYGHTERS_CONTRACT_ADDRESS, FYGHTERS_CONTRACT_ABI, signer);

    const dai = new ethers.Contract(DAI_CONTRACT_ADDRESS, DAI_CONTRACT_ABI, signer);

    const [account] = await ethereum.enable();

    // TODO: It isn't working sometimes
    const { networkVersion: networkId } = ethereum;
    setState({
      metamask: {
        ...metamask,
        contracts: { fyghters, dai },
        ethereum,
        account,
        provider,
        networkId,
        loading: false,
      },
    });
  }
};
