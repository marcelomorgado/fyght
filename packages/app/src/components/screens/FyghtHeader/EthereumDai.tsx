import React from "react";
import { Spin, Button, Typography } from "antd";
import { useFyghtState } from "../../../state";
import { formatDai } from "../../../helpers";
import { MINT_AMOUNT } from "../../../constants";

const { Text } = Typography;

export const EthereumDai: React.FC = () => {
  const [
    {
      daiBalances: {
        ethereumBalance: { loading, amount: balanceInWei },
      },
    },
    { mintDai },
  ] = useFyghtState();

  return (
    <>
      <Text code style={{ color: "#fff" }}>
        {`Dai Balance ${formatDai(balanceInWei)}`}
        {loading ? (
          <>
            {` `}
            <Spin size="small" />
          </>
        ) : null}
      </Text>

      <Button type="primary" onClick={mintDai}>
        {`Mint ${formatDai(MINT_AMOUNT)}`}
      </Button>
    </>
  );
};
