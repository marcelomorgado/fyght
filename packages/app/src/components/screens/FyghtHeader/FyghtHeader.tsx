import React from "react";
import { Layout, Row, Col, Button, Typography } from "antd";
import { useFyghtState } from "../../../state";
import { BigNumber } from "ethers";
import { formatDai } from "../../../helpers";

const { Header } = Layout;
const { Text } = Typography;

type Props = {
  balanceInWei: BigNumber;
};

export const FyghtHeader: React.FC<Props> = ({ balanceInWei }: Props) => {
  const [
    {
      metamask: { ethereum, account },
      balance: { loading },
    },
    { setMetamaskAccount, setErrorMessage, mintDai },
  ] = useFyghtState();

  const onConnect = async (): Promise<void> => {
    if (!ethereum || !ethereum.isMetaMask) {
      setErrorMessage("You have to install Metamask to proceed");
      return;
    }
    const [account] = await ethereum.enable();
    setMetamaskAccount(account);
  };

  return (
    <Header>
      <Row justify={"end"}>
        <Col span={2}>
          <Text code style={{ color: "#fff" }}>
            {`Your Balance ${formatDai(balanceInWei)}`}
          </Text>
        </Col>
        <Col span={2}>
          <Button type="primary" onClick={mintDai} loading={loading}>
            Get Dai
          </Button>
        </Col>
        <Col span={2} offset={18}>
          {!account ? (
            <Button type="primary" onClick={onConnect}>
              Sign in
            </Button>
          ) : null}
        </Col>
      </Row>
    </Header>
  );
};

export default FyghtHeader;
