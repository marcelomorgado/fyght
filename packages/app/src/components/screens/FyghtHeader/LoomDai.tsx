import React from "react";
import { Col, Button, Typography } from "antd";
import { useFyghtState } from "../../../state";
import { formatDai } from "../../../helpers";
import { DEPOSIT_TO_LOOM_AMOUNT } from "../../../constants";

const { Text } = Typography;

export const LoomDai: React.FC = () => {
  const [
    {
      daiBalances: {
        loomBalance: { loading, amount: balanceInWei },
      },
    },
    { depositToLoom, withdrawFromLoom },
  ] = useFyghtState();

  return (
    <>
      <Col span={3}>
        <Text code style={{ color: "#fff" }}>
          {`FyGHT Balance ${formatDai(balanceInWei)}`}
        </Text>
      </Col>
      <Col span={2}>
        <Button type="primary" onClick={depositToLoom} loading={loading}>
          {`Deposit ${formatDai(DEPOSIT_TO_LOOM_AMOUNT)}`}
        </Button>
      </Col>
      <Col span={2}>
        <Button type="primary" onClick={withdrawFromLoom} loading={loading}>
          Withdraw All
        </Button>
      </Col>
    </>
  );
};
