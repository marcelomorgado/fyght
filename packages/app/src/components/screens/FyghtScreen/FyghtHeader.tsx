import React from "react";
import { Layout, Row, Col, Button } from "antd";
import { useFyghtState } from "../../../state";

const { Header } = Layout;

export const FyghtHeader: React.FC = () => {
  const [
    {
      metamask: { ethereum, account },
    },
    { setMetamaskAccount },
  ] = useFyghtState();

  const onConnect = async (): Promise<void> => {
    const [account] = await ethereum.enable();
    setMetamaskAccount(account);
  };

  if (!ethereum || !ethereum.isMetaMask) {
    //return <>{"Please install MetaMask."}</>;
    console.log("Metmask not installed!");
  }

  return (
    <Header>
      <Row justify={"end"}>
        <Col>
          {!account ? (
            <Button type="primary" onClick={onConnect}>
              Connect to metamask
            </Button>
          ) : null}
        </Col>
      </Row>
    </Header>
  );
};

export default FyghtHeader;
