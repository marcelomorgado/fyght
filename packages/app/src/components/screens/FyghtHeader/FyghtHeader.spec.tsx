import * as React from "react";
import { shallow } from "enzyme";
import * as FyghtState from "../../../state";
import { FyghtHeader } from "./FyghtHeader";

describe("FyghtHeader", () => {
  test("should render the component", () => {
    jest
      .spyOn(FyghtState, "useFyghtState")
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .mockImplementation((): any => [
        {
          metamask: { ethereumAccount: "0x", ethereum: { isMetaMask: true } },
        },
        { setMetamaskAccount: jest.fn(), setErrorMessage: jest.fn() },
      ]);

    expect(shallow(<FyghtHeader />)).toMatchSnapshot();
  });
});
