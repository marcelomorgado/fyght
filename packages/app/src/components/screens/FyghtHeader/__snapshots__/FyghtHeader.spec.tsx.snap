// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FyghtScreen should render the component 1`] = `
<Header>
  <Row
    gutter={0}
    justify="end"
  >
    <Col />
  </Row>
</Header>
`;
